azure-iot-device/tests/provisioning/pipeline/test_provisioning_pipeline.py

```diff
@@ -103,9 +87,11 @@ def assert_for_client_x509(x509):
     assert x509.pass_phrase == fake_pass_phrase
 
 
-@pytest.fixture(scope="function")
-def input_security_client(params_security_clients):
-    return params_security_clients["client_class"](**params_security_clients["init_kwargs"])
+@pytest.fixture(scope="function", params=[x509_security_client, symmetric_key_security_client])
+def input_security_client(request):
+    print(request)
+    print(request.param)
+    return request.param
```

Comment:
>request.param [](start = 11, length = 13)

This change has a very subtle but important effect.  

The old fixture created a new security client object for each test, so each test started out with a blank slate.  

This new fixture creates a single global security client (actually one for x509 and one for symmetric key) and uses that one object for all tests.  This means that each test inherits the security client from the previous test (and all tests before), so tests are no longer independent.

This has 2 implications:
1. If a test fails, it could be because of something that a previous test did to affect the state of the security client.  (if test1 sets some security client attribute to an invalid value, test2 will fail if it runs after test 1)
2. If a test passes, it could be because of something that a previous test did to affect the state of the security client.  (If test 1 initializes the security client, and test 2 does not initialize the security client, test2 will fail if it doesn't run after test 1)

Reply:
Yup, the security clients themselves ALSO need to be fixtures

---
In reply to: [392240669](https://github.com/Azure/azure-iot-sdk-python/pull/490#discussion_r392240669) [](ancestors = 392240669)

Reply:
i could not directly use fixtures , so i have to see how to get them created here. 

---
In reply to: [392403164](https://github.com/Azure/azure-iot-sdk-python/pull/490#discussion_r392403164) [](ancestors = 392403164,392240669)